crates/ra_hir/src/adt.rs

```diff
@@ -41,33 +48,37 @@ impl StructData {
     }
 }
 
-impl Enum {
-    pub(crate) fn new(def_id: DefId) -> Self {
-        Enum { def_id }
-    }
+fn get_def_id(
+    db: &impl HirDatabase,
+    same_file_loc: &DefLoc,
+    node: &SyntaxNode,
+    expected_kind: DefKind,
+) -> DefId {
+    let file_id = same_file_loc.source_item_id.file_id;
+    let file_items = db.file_items(file_id);
+
+    let item_id = file_items.id_of(file_id, node);
+    let source_item_id = SourceItemId {
+        item_id: Some(item_id),
+        ..same_file_loc.source_item_id
+    };
+    let loc = DefLoc {
+        kind: expected_kind,
+        source_item_id: source_item_id,
+        ..*same_file_loc
+    };
+    loc.id(db)
 }
 
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct EnumData {
     pub(crate) name: Option<Name>,
-    pub(crate) variants: Vec<(Name, Arc<VariantData>)>,
+    pub(crate) variants: Option<FxHashMap<Name, EnumVariant>>,
```

Comment:
We could also return a `Vec` of pairs here. That may not be optimal when querying for a specific variant tho. The hashmap is more expensive to construct and iterate over OTOH, so it's a trade-off. Guidance here would be appreciated.

Reply:
I think the ideal data structure here would be a map backed by sorted vector, due to three reasons:

* most of enums are small, so even a linear search in a vector should be faster than a hash-map lookup
* for occasional big enums, a log(N) search shouldn't be that much worse than a hash-table lookup
* performance-wise, I am in general much more worried about memory than about time, and sorted vector is as compact as one can get.

Reply:
Sounds good! Is there a data structure that we currently use that does this, or can I just sort the `Vec` and maintain that invariant by hand?

Reply:
No, we don't have implemented this yet. Implementing it "by hand" is probably not a good idea: it's better to search for this on crates.io, or roll our own, but as a separate module/crate. For this PR though, I suggest sticking with the simplest implementation possible.

Tweaking the data structures is better done in a separate PR. I think it makes sense to introduce a `ra_ds` crate which implements/re-exports commonly used data structures, like FxHashMap or SortedVec.

Reply:
:+1: what shall we do with this PR? Just a `Vec` for now?

Reply:
Yeah, let's leave a Vec

Reply:
Done!